### PR TITLE
Update mirror finance strategy for buyback/compound logic

### PR DESCRIPTION
**Pull request by deghub**
Updated mirror finance strategy to accept buyback/compound logic.
Added MIR-UST pair strategy from the mirror finance.

Test results: 

`test/mirror-finance/mir-ust-buyback.js` Block Number `12239121`:
```
  MIR-UST pair reward and buyback test
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0x447f95026107aaed7472A0470931e689f51e0e42
Fetching Underlying at:  0x87dA823B6fC8EB8575a235A824690fda94674c88
New Vault Deployed:  0x660A4a153D1506A44218E41C459DaA897db1b0f5
reward pool needs to be deployed
New PotPool deployed:  0x08eB2d1CeB9d2e43f82812bb118A9b2D2200b824
Strategy Deployed:  0xbaEBDdc34AFA4aF85f7bA2Ef243B033921707200
Strategy and vault added to Controller.
    Happy path
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
iFarm in reward pool:  0
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000368625704200644
growth:  1.0003686257042006
iFarm in reward pool:  1358348
loop  2
old shareprice:  1000368625704200644
new shareprice:  1000737590426305092
growth:  1.0003688287623422
iFarm in reward pool:  2717946
loop  3
old shareprice:  1000737590426305092
new shareprice:  1001106691340991965
growth:  1.0003688288700434
iFarm in reward pool:  4078041
loop  4
old shareprice:  1001106691340991965
new shareprice:  1001475918778949689
growth:  1.0003688192688664
iFarm in reward pool:  5438609
loop  5
old shareprice:  1001475918778949689
new shareprice:  1001845291999868258
growth:  1.0003688288594788
iFarm in reward pool:  6799709
loop  6
old shareprice:  1001845291999868258
new shareprice:  1002214791842724122
growth:  1.0003688192636193
iFarm in reward pool:  8161281
loop  7
old shareprice:  1002214791842724122
new shareprice:  1002584437580362803
growth:  1.0003688288584915
iFarm in reward pool:  9523385
loop  8
old shareprice:  1002584437580362803
new shareprice:  1002954219661872458
growth:  1.0003688288663268
iFarm in reward pool:  10885995
loop  9
old shareprice:  1002954219661872458
new shareprice:  1003324138126719665
growth:  1.000368828863367
iFarm in reward pool:  12249103
farmerNewFarm:     10944218
farmerOldBalance:  152027369121
farmerNewBalance:  152588989033
earned underlying!
earned iFarm!
      ✓ Farmer should earn money (38284ms)


  1 passing (48s)
```

`test/mirror-finance/mamzn-ust-buyback.js` Block Number `12239121`:
```
  MAMZN-UST pair reward and buyback test
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0x447f95026107aaed7472A0470931e689f51e0e42
Fetching Underlying at:  0x0Ae8cB1f57e3b1b7f4f5048743710084AA69E796
New Vault Deployed:  0x660A4a153D1506A44218E41C459DaA897db1b0f5
reward pool needs to be deployed
New PotPool deployed:  0x08eB2d1CeB9d2e43f82812bb118A9b2D2200b824
Strategy Deployed:  0xbaEBDdc34AFA4aF85f7bA2Ef243B033921707200
Strategy and vault added to Controller.
    Happy path
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
iFarm in reward pool:  0
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000415744043490674
growth:  1.0004157440434907
iFarm in reward pool:  150387653
loop  2
old shareprice:  1000415744043490674
new shareprice:  1000831682623373634
growth:  1.0004157657277581
iFarm in reward pool:  300845661
loop  3
old shareprice:  1000831682623373634
new shareprice:  1001247772425830916
growth:  1.0004157440353671
iFarm in reward pool:  451358384
loop  4
old shareprice:  1001247772425830916
new shareprice:  1001664046093004014
growth:  1.0004157548996735
iFarm in reward pool:  601937605
loop  5
old shareprice:  1001664046093004014
new shareprice:  1002080492814390120
growth:  1.0004157548860924
iFarm in reward pool:  752579432
loop  6
old shareprice:  1002080492814390120
new shareprice:  1002497101847702815
growth:  1.00041574408074
iFarm in reward pool:  903279967
loop  7
old shareprice:  1002497101847702815
new shareprice:  1002913894885754444
growth:  1.000415754855833
iFarm in reward pool:  1054047075
loop  8
old shareprice:  1002913894885754444
new shareprice:  1003330861255671759
growth:  1.0004157549038293
iFarm in reward pool:  1204876866
loop  9
old shareprice:  1003330861255671759
new shareprice:  1003748000958651539
growth:  1.0004157548811543
iFarm in reward pool:  1355769365
farmerNewIFarm:     1251790974
farmerOldBalance:  835576305756
farmerNewBalance:  839056752586
earned underlying!
earned iFarm!
      ✓ Farmer should earn money (38409ms)


  1 passing (49s)
```

`test/mirror-finance/mappl-ust-buyback.js` Block Number `12239121`:
```
  MAPPLE-UST pair reward and buyback test
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0x447f95026107aaed7472A0470931e689f51e0e42
Fetching Underlying at:  0xB022e08aDc8bA2dE6bA4fECb59C6D502f66e953B
New Vault Deployed:  0x660A4a153D1506A44218E41C459DaA897db1b0f5
reward pool needs to be deployed
New PotPool deployed:  0x08eB2d1CeB9d2e43f82812bb118A9b2D2200b824
Strategy Deployed:  0xbaEBDdc34AFA4aF85f7bA2Ef243B033921707200
Strategy and vault added to Controller.
    Happy path
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
iFarm in reward pool:  0
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000428299030373423
growth:  1.0004282990303734
iFarm in reward pool:  27743206
loop  2
old shareprice:  1000428299030373423
new shareprice:  1000856803816417619
growth:  1.0004283213364311
iFarm in reward pool:  55499739
loop  3
old shareprice:  1000856803816417619
new shareprice:  1001285480966085742
growth:  1.0004283101718783
iFarm in reward pool:  83267438
@openzeppelin/test-helpers WARN advanceBlockTo: Advancing too many blocks is causing this test to be slow.
loop  4
old shareprice:  1001285480966085742
new shareprice:  1001714330554940404
growth:  1.0004282990186184
iFarm in reward pool:  111046311
loop  5
old shareprice:  1001714330554940404
new shareprice:  1002143375003867243
growth:  1.000428310183692
iFarm in reward pool:  138837802
loop  6
old shareprice:  1002143375003867243
new shareprice:  1002572603217093021
growth:  1.0004283101838838
iFarm in reward pool:  166641193
loop  7
old shareprice:  1002572603217093021
new shareprice:  1003002015271506010
growth:  1.0004283101822602
iFarm in reward pool:  194456493
loop  8
old shareprice:  1003002015271506010
new shareprice:  1003431600056750713
growth:  1.0004282990250308
iFarm in reward pool:  222282984
loop  9
old shareprice:  1003431600056750713
new shareprice:  1003861368836959175
growth:  1.0004282990292355
iFarm in reward pool:  250121392
farmerNewFarm:     230673116
farmerOldBalance:  754341299158
farmerNewBalance:  757578437225
earned underlying!
earned iFarm!
      ✓ Farmer should earn money (46783ms)


  1 passing (58s)
```

`test/mirror-finance/mgoogl-ust-buyback.js` Block Number `12239121`:
```
  MGOOG-UST pair reward and buyback test
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0x447f95026107aaed7472A0470931e689f51e0e42
Fetching Underlying at:  0x4b70ccD1Cf9905BE1FaEd025EADbD3Ab124efe9a
New Vault Deployed:  0x660A4a153D1506A44218E41C459DaA897db1b0f5
reward pool needs to be deployed
New PotPool deployed:  0x08eB2d1CeB9d2e43f82812bb118A9b2D2200b824
Strategy Deployed:  0xbaEBDdc34AFA4aF85f7bA2Ef243B033921707200
Strategy and vault added to Controller.
    Happy path
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
iFarm in reward pool:  0
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000433683832784936
growth:  1.000433683832785
iFarm in reward pool:  66232114
loop  2
old shareprice:  1000433683832784936
new shareprice:  1000867555772161076
growth:  1.0004336838576986
iFarm in reward pool:  132492948
loop  3
old shareprice:  1000867555772161076
new shareprice:  1001301615928893378
growth:  1.0004336839117516
iFarm in reward pool:  198782520
loop  4
old shareprice:  1001301615928893378
new shareprice:  1001735864302981842
growth:  1.0004336838842367
iFarm in reward pool:  265100838
loop  5
old shareprice:  1001735864302981842
new shareprice:  1002170289647167948
growth:  1.0004336725474918
iFarm in reward pool:  331446189
loop  6
old shareprice:  1002170289647167948
new shareprice:  1002604914781340803
growth:  1.0004336839144632
iFarm in reward pool:  397822042
loop  7
old shareprice:  1002604914781340803
new shareprice:  1003039716996376259
growth:  1.0004336725350387
iFarm in reward pool:  464224953
loop  8
old shareprice:  1003039716996376259
new shareprice:  1003474719109855819
growth:  1.0004336838373482
iFarm in reward pool:  530658391
loop  9
old shareprice:  1003474719109855819
new shareprice:  1003909909876828566
growth:  1.000433683837455
iFarm in reward pool:  597120642
farmerNewIFarm:     551166155
farmerOldBalance:  433350046261
farmerNewBalance:  435233082546
earned underlying!
earned iFarm!
      ✓ Farmer should earn money (45206ms)


  1 passing (1m)
```

`test/mirror-finance/mnflx-ust-buyback.js` Block Number `12239121`:
```
  MNFLX-UST pair reward and buyback test
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0x447f95026107aaed7472A0470931e689f51e0e42
Fetching Underlying at:  0xC99A74145682C4b4A6e9fa55d559eb49A6884F75
New Vault Deployed:  0x660A4a153D1506A44218E41C459DaA897db1b0f5
reward pool needs to be deployed
New PotPool deployed:  0x08eB2d1CeB9d2e43f82812bb118A9b2D2200b824
Strategy Deployed:  0xbaEBDdc34AFA4aF85f7bA2Ef243B033921707200
Strategy and vault added to Controller.
    Happy path
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
iFarm in reward pool:  0
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000404172933966685
growth:  1.0004041729339668
iFarm in reward pool:  58559560
loop  2
old shareprice:  1000404172933966685
new shareprice:  1000808519750539981
growth:  1.0004041834565598
iFarm in reward pool:  117144313
loop  3
old shareprice:  1000808519750539981
new shareprice:  1001213030008795892
growth:  1.0004041834679391
iFarm in reward pool:  175752748
loop  4
old shareprice:  1001213030008795892
new shareprice:  1001617703764098396
growth:  1.000404183468626
iFarm in reward pool:  234384874
loop  5
old shareprice:  1001617703764098396
new shareprice:  1002022541073015035
growth:  1.0004041834598123
iFarm in reward pool:  293040699
loop  6
old shareprice:  1002022541073015035
new shareprice:  1002427531465736259
growth:  1.0004041729363569
iFarm in reward pool:  351718704
loop  7
old shareprice:  1002427531465736259
new shareprice:  1002832696106947771
growth:  1.0004041834731126
iFarm in reward pool:  410421952
loop  8
old shareprice:  1002832696106947771
new shareprice:  1003238024497954467
growth:  1.0004041834620871
iFarm in reward pool:  469148926
loop  9
old shareprice:  1003238024497954467
new shareprice:  1003643516723005879
growth:  1.000404183469077
iFarm in reward pool:  527899632
farmerNewIFarm:     487267251
farmerOldBalance:  830865161366
farmerNewBalance:  834229478010
earned underlying!
earned!
      ✓ Farmer should earn money (45292ms)


  1 passing (60s)
```

`test/mirror-finance/mtsla-ust-buyback.js` Block Number `12239121`:
```
  MTSLA-UST pair reward and buyback test
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0x447f95026107aaed7472A0470931e689f51e0e42
Fetching Underlying at:  0x5233349957586A8207c52693A959483F9aeAA50C
New Vault Deployed:  0x660A4a153D1506A44218E41C459DaA897db1b0f5
reward pool needs to be deployed
New PotPool deployed:  0x08eB2d1CeB9d2e43f82812bb118A9b2D2200b824
Strategy Deployed:  0xbaEBDdc34AFA4aF85f7bA2Ef243B033921707200
Strategy and vault added to Controller.
    Happy path
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
iFarm in reward pool:  0
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000346359261485854
growth:  1.0003463592614859
iFarm in reward pool:  13161720
loop  2
old shareprice:  1000346359261485854
new shareprice:  1000692856467916956
growth:  1.0003463772354677
iFarm in reward pool:  26328689
@openzeppelin/test-helpers WARN advanceBlockTo: Advancing too many blocks is causing this test to be slow.
loop  3
old shareprice:  1000692856467916956
new shareprice:  1001039464644324864
growth:  1.0003463681929652
iFarm in reward pool:  39499870
loop  4
old shareprice:  1001039464644324864
new shareprice:  1001386192882731032
growth:  1.0003463682009073
iFarm in reward pool:  52675617
loop  5
old shareprice:  1001386192882731032
new shareprice:  1001733041183135461
growth:  1.0003463681673161
iFarm in reward pool:  65855929
loop  6
old shareprice:  1001733041183135461
new shareprice:  1002080009696086110
growth:  1.0003463682425218
iFarm in reward pool:  79040807
loop  7
old shareprice:  1002080009696086110
new shareprice:  1002427098416206267
growth:  1.0003463682707585
iFarm in reward pool:  92230251
loop  8
old shareprice:  1002427098416206267
new shareprice:  1002774298256851191
growth:  1.0003463591927968
iFarm in reward pool:  105423916
loop  9
old shareprice:  1002774298256851191
new shareprice:  1003121627391310364
growth:  1.0003463682057496
iFarm in reward pool:  118622501
farmerNewIFarm:     109327507
farmerOldBalance:  185987242621
farmerNewBalance:  186632448355
earned underlying!
earned!
      ✓ Farmer should earn money (44960ms)


  1 passing (55s)
```

`test/mirror-finance/mtwtr-ust-buyback.js` Block Number `12239121`:
```
  MTWTR-UST pair reward and buyback test
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0xB4865AA43A0F32ede7e2Fcb5bcCe1190da5De72E
Fetching Underlying at:  0x34856be886A2dBa5F7c38c4df7FD86869aB08040
New Vault Deployed:  0x660A4a153D1506A44218E41C459DaA897db1b0f5
reward pool needs to be deployed
New PotPool deployed:  0x08eB2d1CeB9d2e43f82812bb118A9b2D2200b824
Strategy Deployed:  0xbaEBDdc34AFA4aF85f7bA2Ef243B033921707200
Strategy and vault added to Controller.
    Happy path
loop  0
old shareprice:  1000000000000000000
new shareprice:  1000000000000000000
growth:  1
iFarm in reward pool:  0
loop  1
old shareprice:  1000000000000000000
new shareprice:  1000374952589042107
growth:  1.000374952589042
iFarm in reward pool:  65483438808831
loop  2
old shareprice:  1000374952589042107
new shareprice:  1000750055534730327
growth:  1.0003749623525835
iFarm in reward pool:  130993134024623
loop  3
old shareprice:  1000750055534730327
new shareprice:  1001125279586386647
growth:  1.000374942823716
iFarm in reward pool:  196523977172360
loop  4
old shareprice:  1001125279586386647
new shareprice:  1001500654100135910
growth:  1.000374952587257
iFarm in reward pool:  262081095137004
loop  5
old shareprice:  1001500654100135910
new shareprice:  1001876169360933900
growth:  1.0003749525866614
iFarm in reward pool:  327662791235919
loop  6
old shareprice:  1001876169360933900
new shareprice:  1002251825421553422
growth:  1.0003749525860657
iFarm in reward pool:  393269074682700
loop  7
old shareprice:  1002251825421553422
new shareprice:  1002627612548663962
growth:  1.0003749428213338
iFarm in reward pool:  458898245601447
loop  8
old shareprice:  1002627612548663962
new shareprice:  1003003550363654777
growth:  1.0003749525848737
iFarm in reward pool:  524553730757780
loop  9
old shareprice:  1003003550363654777
new shareprice:  1003379629136902893
growth:  1.0003749525842773
iFarm in reward pool:  590233830923229
farmerNewIFarm:     545107184965146
farmerOldBalance:  2773920714016238025
farmerNewBalance:  2784339168314113455
earned underlying!
earned!
      ✓ Farmer should earn money (43836ms)


  1 passing (55s)
```

### CODE REVIEW

File: test/mirror-finance/mamzn-ust-buyback.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+// Vanilla Mocha test. Increased compatibility with tools that integrate Mocha.
+describe("MAMZN-UST pair reward and buyback test", function() {
+  let accounts;
+
+  // external contracts
+  let underlying;
+
+  // external setup
+  let underlyingWhale = "0x447f95026107aaed7472A0470931e689f51e0e42";
```
Comment (CryptJS13) [inline test/mirror-finance/mamzn-ust-buyback.js:21]: Please include the blockNumber you used in a comment to allow reproduction of your tests

File: contracts/base/snx-base/SNXRewardStrategyWithBuyback.sol [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  *   Resumes the ability to invest into the underlying reward pools
+  */
+
+  function continueInvesting() public onlyGovernance {
+    pausedInvesting = false;
+  }
+
+
+  function setLiquidationPaths(address [] memory _uniswapRouteFarm) public onlyGovernance {
+    uniswapRoutes[farm] = _uniswapRouteFarm;
```
[the remark this answers is not in the document]
Comment (deghub) [reply contracts/base/snx-base/SNXRewardStrategyWithBuyback.sol:123]: resolve it

File: contracts/strategies/mirror-finance/MirrorMainnet_mAAPL_UST.sol [hunk 3]
```diff
@@ -26,4 +26,4 @@ contract MirrorMainnet_mAAPL_UST is SNXReward2FarmStrategy {
     // adding ability to liquidate reward tokens manually if there is no liquidity
     unsalvagableTokens[mir] = false;
   }
-}
```
[the remark this answers is not in the document]
Comment (deghub) [reply contracts/strategies/mirror-finance/MirrorMainnet_mAAPL_UST.sol]: did it

File: contracts/strategies/mirror-finance/MirrorV2Mainnet_MIR_UST.sol [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  constructor(
+    address _storage,
+    address _vault,
+    address _distributionPool
+  )
+  SNXRewardStrategyWithBuyback(_storage, mir_ust, _vault, rewardPool, mir, uniswapRouterAddress, farm, _distributionPool, 5000)
+  public {
+    require(IVault(_vault).underlying() == mir_ust, "Underlying mismatch");
+    uniswapRoutes[farm] = [mir, weth, farm];
```
[the remark this answers is not in the document]
Comment (deghub) [reply contracts/strategies/mirror-finance/MirrorV2Mainnet_MIR_UST.sol:24]: right, removed it

File: contracts/strategies/mirror-finance/MirrorV2Mainnet_MIR_UST.sol [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    uniswapRoutes[ust] = [mir, ust];
+
+    // adding ability to liquidate reward tokens manually if there is no liquidity
+    unsalvagableTokens[mir] = false;
```
[the remark this answers is not in the document]
Comment (deghub) [reply contracts/strategies/mirror-finance/MirrorV2Mainnet_MIR_UST.sol:28]: removed

File: test/1inch/eth-dai.js [hunk 6]
```diff
@@ -44,6 +44,7 @@ describe("Mainnet ETH/DAI", function() {
     await send.ether(etherGiver, underlyingWhale, "1" + "000000000000000000");
 
     farmerBalance = await underlying.balanceOf(underlyingWhale);
+    console.log('farmerBalance: \n', farmerBalance.toString());
     await underlying.transfer(farmer1, farmerBalance, { from: underlyingWhale });
```
[the remark this answers is not in the document]
Comment (deghub) [reply test/1inch/eth-dai.js]: yep, removed it

**Pull request merged**